### PR TITLE
Bump to k8s 1.31

### PR DESCRIPTION
**Pull request by ArangoGutierrez**
I upgraded the K8s library versions to v1.31, and controller-tools to v0.16.4.

Fixes: #652

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -33,8 +33,8 @@ KUBEBUILDER_ASSETS_PATH := $(dir $(abspath $(firstword $(MAKEFILE_LIST))))bin/ku
 KIND_VERSION=v0.18.0
 HELM_VERSION=v3.11.2
 # This kubectl version supports -k for kustomization.
-KUBECTL_VERSION=v1.30.4
-ENVTEST_K8S_VERSION=1.30.0
+KUBECTL_VERSION=v1.31.1
+ENVTEST_K8S_VERSION=1.31.1
```
Comment (tenzen-y) [inline Makefile:37]:
```suggestion
ENVTEST_K8S_VERSION=1.31.0
```

It seems that the latest version is v1.30.0:

```
setup-envtest list
(installed)  v1.31.0  darwin/arm64
(installed)  v1.30.0  darwin/arm64
(installed)  v1.29.3  darwin/arm64
(installed)  v1.29.1  darwin/arm64
(installed)  v1.28.3  darwin/arm64
(installed)  v1.27.1  darwin/arm64
(available)  v1.30.0  darwin/arm64
(available)  v1.29.3  darwin/arm64
(available)  v1.29.1  darwin/arm64
(available)  v1.29.0  darwin/arm64
(available)  v1.28.3  darwin/arm64
(available)  v1.28.0  darwin/arm64
(available)  v1.27.1  darwin/arm64
(available)  v1.26.1  darwin/arm64
(available)  v1.26.0  darwin/arm64
(available)  v1.25.0  darwin/arm64
(available)  v1.24.2  darwin/arm64
(available)  v1.24.1  darwin/arm64
```

File: go.mod [hunk 2]
```diff
@@ -1,18 +1,18 @@
 module github.com/kubeflow/mpi-operator
 
-go 1.22.0
+go 1.23.0
```
Comment (tenzen-y) [inline go.mod:3]:
```suggestion
go 1.23
```

Could we eliminate the patch version?

**Review by tenzen-y — COMMENTED** [review]
Thanks!

File: Makefile (hunk 1 above)
Comment (ArangoGutierrez) [reply Makefile]: done

File: go.mod (hunk 2 above)
Comment (ArangoGutierrez) [reply go.mod]: done

**Review by tenzen-y — COMMENTED** [review]
Thank you!
/lgtm
/approve

/hold for CI

**Comment by tenzen-y** [comment]
/lgtm cancel

We need to upgrade the golangci-lint version.

**Comment by tenzen-y** [comment]
I chat with @ArangoGutierrez offline.
He is trying to resolve CI errors now.

File: pkg/controller/mpi_job_controller.go [hunk 3]
```diff
@@ -339,7 +339,7 @@ func NewMPIJobControllerWithClock(
 		priorityClassSynced: priorityClassSynced,
 		mpiJobLister:        mpiJobInformer.Lister(),
 		mpiJobSynced:        mpiJobInformer.Informer().HasSynced,
-		queue:               workqueue.NewRateLimitingQueueWithConfig(workqueue.DefaultControllerRateLimiter(), workqueue.RateLimitingQueueConfig{Name: "MPIJobs"}),
+		queue:               workqueue.NewTypedRateLimitingQueueWithConfig(workqueue.DefaultTypedControllerRateLimiter[any](), workqueue.TypedRateLimitingQueueConfig[any]{Name: "MPIJob"}),
```
Comment (tenzen-y) [inline pkg/controller/mpi_job_controller.go:342]: I guess that we need to do this:

```suggestion
		queue:               workqueue.NewTypedRateLimitingQueueWithConfig(workqueue.DefaultTypedControllerRateLimiter[string](), workqueue.TypedRateLimitingQueueConfig[string]{Name: "MPIJob"}),
```

and then,

https://github.com/kubeflow/mpi-operator/blob/dbcbfae1f44d0dd9c4f99ed85908f24f19818805/pkg/controller/mpi_job_controller.go#L250

```golang
workqueue.TypedRateLimitingInterface[string]
````

Comment (tenzen-y) [reply pkg/controller/mpi_job_controller.go:342]: Modified my comment related to types.

Comment (ArangoGutierrez) [reply pkg/controller/mpi_job_controller.go:342]: yeah I was about to comment that `any` is required.

**Comment by tenzen-y** [comment]
Additionally, I guess that we need to specify the 22.04: 

- https://github.com/kubeflow/mpi-operator/blob/86d09e343524a09faa1a1892876c3850ccfd1148/.github/workflows/main.yml#L6
- https://github.com/kubeflow/mpi-operator/blob/86d09e343524a09faa1a1892876c3850ccfd1148/.github/workflows/mpi-operator-docker-image-publish.yml#L18
- https://github.com/kubeflow/mpi-operator/blob/86d09e343524a09faa1a1892876c3850ccfd1148/.github/workflows/golangci-lint.yml#L12
-

**Comment by tenzen-y** [comment]
It seems that the latest ubuntu runner has many issues: https://github.com/actions/runner-images/issues?q=is%3Aissue+Ubuntu+24.04+runner

File: pkg/controller/mpi_job_controller.go (hunk 3 above)
Comment (tenzen-y) [reply pkg/controller/mpi_job_controller.go:342]: @ArangoGutierrez Could you address the Line 250 as I mentioned above?

Comment (ArangoGutierrez) [reply pkg/controller/mpi_job_controller.go:342]: I did

**Comment by tenzen-y** [comment]
Could you try to change the installation URL in the following?

https://github.com/kubeflow/mpi-operator/blob/86d09e343524a09faa1a1892876c3850ccfd1148/Makefile#L149

```
curl -L -o $(PROJECT_DIR)/bin/kubectl https://dl.k8s.io/release/${KUBECTL_VERSION}/bin/$(GOOS)/$(GOARCH)/kubectl 
```

File: test/e2e/e2e_suite_test.go [hunk 4]
```diff
@@ -61,7 +61,7 @@ const (
 	volcanoSchedulerManifestPath   = rootPath + "/dep-manifests/volcano-scheduler/" // all in one yaml of volcano-development.yaml
 	envUseExistingSchedulerPlugins = "USE_EXISTING_SCHEDULER_PLUGINS"
 	envUseExistingVolcanoScheduler = "USE_EXISTING_VOLCANO_SCHEDULER"
-	defaultSchedulerPluginsVersion = "v0.29.8"
+	defaultSchedulerPluginsVersion = "v0.31.1"
```
Comment (tenzen-y) [inline test/e2e/e2e_suite_test.go:64]:
```suggestion
	defaultSchedulerPluginsVersion = "v0.29.8"
```

This is the scheduler-plugins version, not k8s version.

Comment (ArangoGutierrez) [reply test/e2e/e2e_suite_test.go]: changed back

**Review by tenzen-y — COMMENTED** [review]
Thank you!
This was not trival efforts!

/lgtm
/approve

**Comment by ArangoGutierrez** [comment]
> Thank you!
> 
> /lgtm
> 
> /approve
> 
> 
> 
> /hold for CI

Should we remove the hold?

**Comment by tenzen-y** [comment]
Yup
/hold cancel

**Pull request merged**